docs/man/gearmand.8
.TH "GEARMAND" "8" "September 14, 2012" "0.38" "Gearmand"
.SH NAME
gearmand \- Gearmand Documentation, http://gearman.info/
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.SH SYNOPSIS
.sp
\fBGeneral options\fP
.INDENT 0.0
.TP
.B \-b [ \-\-backlog ] arg (=32)
Number of backlog connections for listen.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-check\-args
Check command line and configuration file argments and then exit.
.UNINDENT
.INDENT 0.0
.TP
.B \-d [ \-\-daemon ]
Daemon, detach and run in the background.
.UNINDENT
.INDENT 0.0
.TP
.B \-f [ \-\-file\-descriptors ] arg
Number of file descriptors to allow for the process (total connections will be slightly less). Default is max allowed for user.
.UNINDENT
.INDENT 0.0
.TP
.B \-h [ \-\-help ]
Print this help menu.
.UNINDENT
.INDENT 0.0
.TP
.B \-j [ \-\-job\-retries ] arg (=0)
Number of attempts to run the job before the job server removes it. This is helpful to ensure a bad job does not crash all available workers. Default is no limit.
.UNINDENT
.INDENT 0.0
.TP
.B \-l [ \-\-log\-file ] arg
Log file to write errors and information to.  Turning this option on also forces the first verbose level to be enabled.
.UNINDENT
.INDENT 0.0
.TP
.B \-L [ \-\-listen ] arg
Address the server should listen on. Default is INADDR_ANY.
.UNINDENT
.INDENT 0.0
.TP
.B \-p [ \-\-port ] arg (=4730)
Port the server should listen on.
.UNINDENT
.INDENT 0.0
.TP
.B \-P [ \-\-pid\-file ] arg
File to write process ID out to.
.UNINDENT
.INDENT 0.0
.TP
.B \-r [ \-\-protocol ] arg
Load protocol module.
.UNINDENT
.INDENT 0.0
.TP
.B \-R [ \-\-round\-robin ]
Assign work in round\-robin order per worker connection. The default is to assign work in the order of functions added by the worker.
.UNINDENT
.INDENT 0.0
.TP
.B \-q [ \-\-queue\-type ] arg
Persistent queue type to use.
.UNINDENT
.INDENT 0.0
.TP
.B \-t [ \-\-threads ] arg (=4)
Number of I/O threads to use. Default=4.
.UNINDENT
.INDENT 0.0
.TP
.B \-u [ \-\-user ] arg
Switch to given user after startup.
.UNINDENT
.INDENT 0.0
.TP
.B \-v [ \-\-verbose ] arg (=v)
Increase verbosity level by one.
.UNINDENT
.INDENT 0.0
.TP
.B \-V [ \-\-version ]
Display the version of gearmand and exit.
.UNINDENT
.INDENT 0.0
.TP
.B \-w [ \-\-worker\-wakeup ] arg (=0)
Number of workers to wakeup for each job received. The default is to wakeup all available workers.
.UNINDENT
.sp
\fBHTTP:\fP
.INDENT 0.0
.TP
.B \-\-http\-port arg (=8080)
Port to listen on.
.UNINDENT
.sp
\fBsqlite\fP
.INDENT 0.0
.TP
.B \-\-libsqlite3\-db arg
Database file to use.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libsqlite3\-table arg (=gearman_queue)
Table to use.
.UNINDENT
.sp
\fBMemcached(libmemcached)\fP
.INDENT 0.0
.TP
.B \-\-libmemcached\-servers arg
List of Memcached servers to use.
.UNINDENT
.sp
\fBDrizzle/MySQL(libdrizzle)\fP
.INDENT 0.0
.TP
.B \-host arg
Host of server.
.UNINDENT
.INDENT 0.0
.TP
.B \-port arg
Port of server. (by default Drizzle)
.UNINDENT
.INDENT 0.0
.TP
.B \-uds arg
Unix domain socket for server.
.UNINDENT
.INDENT 0.0
.TP
.B \-user arg
User name for authentication.
.UNINDENT
.INDENT 0.0
.TP
.B \-password arg
Password for authentication.
.UNINDENT
.INDENT 0.0
.TP
.B \-db arg
Schema/Database to use.
.UNINDENT
.INDENT 0.0
.TP
.B \-table arg
Table to use.
.UNINDENT
.INDENT 0.0
.TP
.B \-mysql arg
Use MySQL protocol.
.UNINDENT
.sp
\fBPostgres\fP
.INDENT 0.0
.TP
.B \-\-libpq\-conninfo arg
PostgreSQL connection information string.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libpq\-table arg (=queue)
Table to use.
.UNINDENT
.sp
\fBtokyocabinet\fP
.INDENT 0.0
.TP
.B \-\-libtokyocabinet\-file arg
File name of the database. [see: man tcadb, tcadbopen() for name guidelines]
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libtokyocabinet\-optimize
Optimize database on open. [default=true]
.UNINDENT
.SH DESCRIPTION
.sp
Gearman provides a generic application framework to farm out work to other machines or processes that are better suited to do the work. It allows you to do work in parallel, to load balance processing, and to call functions between languages. It can be used in a variety of applications, from high\-availability web sites to the transport of database replication events. In other words, it is the nervous system for how distributed processing communicates. A few strong points about Gearman:
.INDENT 0.0
.IP \(bu 2
Open Source \- It\(aqs free! (in both meanings of the word) Gearman has an active open source community that is easy to get involved with if you need help or want to contribute.
.IP \(bu 2
Multi\-language \- There are interfaces for a number of languages, and this list is growing. You also have the option to write heterogeneous applications with clients submitting work in one language and workers performing that work in another.
.IP \(bu 2
Flexible \- You are not tied to any specific design pattern. You can quickly put together distributed applications using any model you choose, one of those options being Map/Reduce.
.IP \(bu 2
Fast \- Gearman has a simple protocol and interface with a new optimized server in C to minimize your application overhead.
.IP \(bu 2
Embeddable \- Since Gearman is fast and lightweight, it is great for applications of all sizes. It is also easy to introduce into existing applications with minimal overhead.
.IP \(bu 2
No single point of failure \- Gearman can not only help scale systems, but can do it in a fault tolerant way.
.UNINDENT
.SS Thread Model
.sp
The \-t option to gearmand allows you to specify multiple I/O threads, this is enabled by default. There are currently three types of threads in the job server:
.sp
Listening and management thread \- only one
I/O thread \- can have many
Processing thread \- only one
.sp
When no \-t option is given or \-t 0 is given, all of three thread types happen within a single thread. When \-t 1 is given, there is a thread for listening/management and a thread for I/O and processing. When \-t 2 is given, there is a thread for each type of thread above. For all \-t option values above 2, more I/O threads are created.
.sp
The listening and management thread is mainly responsible for accepting new connections and assigning those connections to an I/O thread (if there are many). It also coordinates startup and shutdown within the server. This thread will have an instance of libevent for managing socket events and signals on an internal pipe. This pipe is used to wakeup the thread or to coordinate shutdown.
.sp
The I/O thread is responsible for doing the read and write system calls on the sockets and initial packet parsing. Once the packet has been parsed it it put into an asynchronous queue for the processing thread (each thread has it\(aqs own queue so there is very little contention). Each I/O thread has it\(aqs own instance of libevent for managing socket events and signals on an internal pipe like the listening thread.
.sp
The processing thread should have no system calls within it (except for the occasional brk() for more memory), and manages the various lists and hash tables used for tracking unique keys, job handles, functions, and job queues. All packets that need to be sent back to connections are put into an asynchronous queue for the I/O thread. The I/O thread will pick these up and send them back over the connected socket. All packets flow through the processing thread since it contains the information needed to process the packets. This is due to the complex nature of the various lists and hash tables. If multiple threads were modifying them the locking overhead would most likely cause worse performance than having it in a single thread (and would also complicate the code). In the future more work may be pushed to the I/O threads, and the processing thread can retain minimal functionality to manage those tables and lists. So far this has not been a significant bottleneck, a 16 core Intel machine is able to process upwards of 50k jobs per second.
.sp
For thread safety to work when UUID are generated, you must be running the uuidd daemon.
.SS Persistent Queues
.sp
Inside the Gearman job server, all job queues are stored in memory. This means if a server restarts or crashes with pending jobs, they will be lost and are never run by a worker. Persistent queues were added to allow background jobs to be stored in an external durable queue so they may live between server restarts and crashes. The persistent queue is only enabled for background jobs because foreground jobs have an attached client. If a job server goes away, the client can detect this and restart the foreground job somewhere else (or report an error back to the original caller). Background jobs on the other hand have no attached client and are simply expected to be run when submitted.
.sp

ChangeLog
1.1.1 Wed Sep 19 22:04:56 EDT 2012
* Merge with 0.39

1.1.0 Wed Sep  5 08:33:37 PDT 2012
* Fix for ABI compatibility issues.

0.39 Wed Sep 19 03:36:06 EDT 2012
* Added gearman_job_error() interface.
* gearman_client_errno() and gearman_worker_errno() now return EINVAL if client or worker is invalid.

0.38 Fri Sep 14 02:16:01 EDT 2012
* Updates for compiling on different platforms.
* UUID portability fixes.

0.37 Sun Sep  2 22:51:08 EDT 2012
* Fix compiler issues around older compilers. Bug 1039629.

0.36 Sat Aug 18 16:57:13 EDT 2012
* Fix issue where sqlite was not correctly being shutdown.
* Check for more warnings (and cleanup found from).
* Fix thread safety issue for uuid_generate().

0.35 Sun Aug  5 16:03:21 EDT 2012
* Critical bug fix for rare case in which initialization of pipe for a worker fails.
* Add support for MySQL specific driver to specify port.
* Fix library issue with OSX Mountain Lion.

0.34 Tue Jul 10 21:19:10 EDT 2012
* Alpha support for direct MySQL queue.
* XML option for testing under valgrind.
* gearmand now allows for a user to specify the format for the server key
* First pass support for checking status via unique value.

0.33 Thu May  3 22:09:29 PDT 2012
* Add --config-file option
* Additional testing for Drizzle
* Fix for OSX 10.6.8 compile issue.

0.32 Wed Apr 11 02:25:38 PDT 2012
* Fix for shutdown bug.

0.31 Sat Apr  7 11:56:58 PDT 2012
* Fixes for bugs 975591, 974712, 970828, 967859, 967463,
* Valgrind fix around epoch retry.
* Numerous issues cleaned up around bouncing server during testing.

0.29 Tue Mar 13 22:58:16 PDT 2012
* Additional Fixes for Postgres
* Additional debug calls in server via tuneable logging.
* Documentation update.
* Fixes for Ubuntu 11.10

0.28 Thu Jan 26 22:43:46 PST 2012
* -v has been deprecated because of parsing issues. Please use --verbose instead
* --log-file=stderr will set output to stderr.
* Fix for gearman.spec file for rpm build.

0.27 Mon Jan  9 20:02:16 PST 2012
* Fix for TokyoCabinet
* Build updates for building on Ubuntu

0.26 Wed Nov 30 09:25:49 PST 2011
* Completed support for syslog for gearmand
* Fix libdrizzle issues
* Add API for killing a worker.

0.25 Thu Nov  3 09:18:46 PDT 2011
  * 1.0 libgearman API extracted.
  * Fix for long function names.
  * Fix for Worker consuming CPU by hanging.
  * TokyoCabinet build fix.
  * Fix for 32bit fix.

0.24 Sun Jul 24 20:57:02 PDT 2011
  * Minor bug fixes.
  * Updated test framework.
  * make fixes for Fedora.

0.23 Wed Jun 29 13:26:26 PDT 2011
  * Defined workers can now return GEARMAN_SHUTDOWN.
  * Benchmark worker can now be told to shutdown.
  * Allocator code has been cleaned up (gearman_allocator_t).
  * Added "workers" option to gearadmin
  * Workers will now default to -1 as timeout (lowers CPU on gearmand server for non-active workers).
  * SO_KEEPALIVE is now enabled on client/worker connections.
  * By default, workers now grab the unique value of the job.

0.22 Sun Jun 19 20:32:18 PDT 2011
  * Added gearman_work_map() for gearman_work_t describer.
  * gearman_execute_by_partition() added. 
  * gearamn_task_attr_t added.
  * Additional documentation update.
  * Issue with PHP client using automated task free (Thanks Taneli Leppä!!)
  * Fixed issue where PHP client's custom allocator was free'ing memory that did not belong to it.

0.21 Wed Jun 15 21:01:04 PDT 2011
  * Support for "drop function" in admin client
  * Fixed issue where server would not report on a bad unique value due to size
  * Forces do operations to complete no matter how the worker is designed.
  * Tasks inherit actions from client, and use them throughout execution.
  * Support for namespaces.
  * gearman_worker_define_function() has been added. It allows jobs to map out work to other workers and then aggregate the results.
  * gearman_client_run_tasks() will now reset the IO state for all connections on exit if all hosts have been marked down. The next call will check all hosts.
  * Lots of bug fixes.

0.20 Fri Apr  8 09:25:15 PDT 2011
  * gearman_client_execute() interface added
  * New types of gearman_unique_t, gearman_function_st, gearman_workload_t
  * Better error reporting from clients.
  * Server no longer hangs because of bad job list.
  * First bit of work on supporting epoch portion of protocol. John Ewart, john@johnewart.net 
  * Improved error message handling in client. 
  * Updates for manpages.

0.19 Fri Apr  1 09:50:22 PDT 2011
  * Fix bug where gearman util would create duplicate output.
  * Update sqlite queue so that it uses a prepared statement.
  * Fixes a number cases where server was not initializing default values internally correctly.  
  * gearman_client_compare()  API added. Returns true if the client connections are connects are equivalent.

0.18 Tue Mar 15 13:32:33 PDT 2011
  * Bug fix for compile when no plugins are requested.
  * Bug fix where during startup using Drizzle or MySQL the server could crash.
  * Bug fix to make sure test programs will run if NULL sized objects are sent.


0.17 Tue Mar  8 08:27:04 PST 2011

gearmand/gearmand.c
/* Gearman server and library
 * Copyright (C) 2008 Brian Aker, Eric Day
 * All rights reserved.
 *
 * Use and distribution licensed under the BSD license.  See
 * the COPYING file in the parent directory for full text.
 */

#include "config.h"

#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#endif
#ifdef HAVE_PWD_H
#include <pwd.h>
#endif
#ifdef HAVE_SIGNAL_H
#include <signal.h>
#endif
#ifdef HAVE_STDIO_H
#include <stdio.h>
#endif
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#ifdef HAVE_SYS_RESOURCE_H
#include <sys/resource.h>
#endif
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifdef TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# ifdef HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

#include <libgearman-server/gearmand.h>

#ifdef HAVE_LIBDRIZZLE
#include <libgearman-server/queue_libdrizzle.h>
#endif

#ifdef HAVE_LIBMEMCACHED
#include <libgearman-server/queue_libmemcached.h>
#endif

#ifdef HAVE_LIBSQLITE3
#include <libgearman-server/queue_libsqlite3.h>
#endif

#ifdef HAVE_LIBPQ
#include <libgearman-server/queue_libpq.h>
#endif

#ifdef HAVE_LIBTOKYOCABINET
#include <libgearman-server/queue_libtokyocabinet.h>
#endif

#include <libgearman-server/protocol_http.h>

#define GEARMAND_LOG_REOPEN_TIME 60
#define GEARMAND_LISTEN_BACKLOG 32

typedef struct
{
  const char *file;
  int fd;
  time_t reopen;
} gearmand_log_info_st;

static gearmand_st *_gearmand;

static bool _set_fdlimit(rlim_t fds);
static bool _pid_write(const char *pid_file);
static void _pid_delete(const char *pid_file);
static bool _switch_user(const char *user);
static bool _set_signals(void);
static void _shutdown_handler(int signal_arg);
static void _log(const char *line, gearman_verbose_t verbose, void *context);

int main(int argc, char *argv[])
{
  gearman_conf_st conf;
  gearman_conf_module_st module;
  const char *name;
  const char *value;
  int backlog= GEARMAND_LISTEN_BACKLOG;
  rlim_t fds= 0;
  uint8_t job_retries= 0;
  uint8_t worker_wakeup= 0;
  in_port_t port= 0;
  const char *host= NULL;
  const char *pid_file= NULL;
  const char *queue_type= NULL;
  uint32_t threads= 0;
  const char *user= NULL;
  uint8_t verbose= 0;
  gearman_return_t ret;
  gearmand_log_info_st log_info;
  bool close_stdio= false;
  int fd;
  bool round_robin= false;

  log_info.file= NULL;
  log_info.fd= -1;
  log_info.reopen= 0;

  if (gearman_conf_create(&conf) == NULL)
  {
    fprintf(stderr, "gearmand: gearman_conf_create: NULL\n");
    return 1;
  }

  if (gearman_conf_module_create(&conf, &module, NULL) == NULL)
  {
    fprintf(stderr, "gearmand: gearman_conf_module_create: NULL\n");
    return 1;
  }

  /* Add all main configuration options. */
#define MCO(__name, __short, __value, __help) \
  gearman_conf_module_add_option(&module, __name, __short, __value, __help);

  MCO("backlog", 'b', "BACKLOG", "Number of backlog connections for listen.")
  MCO("daemon", 'd', NULL, "Daemon, detach and run in the background.")
  MCO("file-descriptors", 'f', "FDS",
      "Number of file descriptors to allow for the process (total connections "
      "will be slightly less). Default is max allowed for user.")
  MCO("help", 'h', NULL, "Print this help menu.");
  MCO("job-retries", 'j', "RETRIES",
      "Number of attempts to run the job before the job server removes it. This"
      "is helpful to ensure a bad job does not crash all available workers. "
      "Default is no limit.")
  MCO("log-file", 'l', "FILE",
      "Log file to write errors and information to. Turning this option on "
      "also forces the first verbose level to be enabled.")
  MCO("listen", 'L', "ADDRESS",
      "Address the server should listen on. Default is INADDR_ANY.")
  MCO("port", 'p', "PORT", "Port the server should listen on.")
  MCO("pid-file", 'P', "FILE", "File to write process ID out to.")
  MCO("protocol", 'r', "PROTOCOL", "Load protocol module.")
  MCO("round-robin", 'R', NULL, "Assign work in round-robin order per worker"
      "connection. The default is to assign work in the order of functions "
      "added by the worker.")
  MCO("queue-type", 'q', "QUEUE", "Persistent queue type to use.")
  MCO("threads", 't', "THREADS", "Number of I/O threads to use. Default=0.")
  MCO("user", 'u', "USER", "Switch to given user after startup.")
  MCO("verbose", 'v', NULL, "Increase verbosity level by one.")
  MCO("version", 'V', NULL, "Display the version of gearmand and exit.")
  MCO("worker-wakeup", 'w', "WORKERS",
      "Number of workers to wakeup for each job received. The default is to "
      "wakeup all available workers.")

  /* Make sure none of the gearman_conf_module_add_option calls failed. */
  if (gearman_conf_return(&conf) != GEARMAN_SUCCESS)
  {
    fprintf(stderr, "gearmand: gearman_conf_module_add_option: %s\n",
            gearman_conf_error(&conf));
    return 1;
  }

  /* Add queue configuration options. */

#ifdef HAVE_LIBDRIZZLE
  if (gearman_server_queue_libdrizzle_conf(&conf) != GEARMAN_SUCCESS)
  {
    fprintf(stderr, "gearmand: gearman_queue_libdrizzle_conf: %s\n",
            gearman_conf_error(&conf));
    return 1;
  }
#endif

#ifdef HAVE_LIBMEMCACHED
  if (gearman_server_queue_libmemcached_conf(&conf) != GEARMAN_SUCCESS)
  {
    fprintf(stderr, "gearmand: gearman_queue_libmemcached_conf: %s\n",
            gearman_conf_error(&conf));
    return 1;
  }
#endif
#ifdef HAVE_LIBTOKYOCABINET
  if (gearman_server_queue_libtokyocabinet_conf(&conf) != GEARMAN_SUCCESS)
  {
    fprintf(stderr, "gearmand: gearman_queue_libtokyocabinet_conf: %s\n",
            gearman_conf_error(&conf));
    return 1;
  }
#endif

#ifdef HAVE_LIBSQLITE3
  if (gearman_server_queue_libsqlite3_conf(&conf) != GEARMAN_SUCCESS)
  {
    fprintf(stderr, "gearmand: gearman_queue_libsqlite3_conf: %s\n",
            gearman_conf_error(&conf));
    return 1;
  }
#endif

#ifdef HAVE_LIBPQ
  if (gearman_server_queue_libpq_conf(&conf) != GEARMAN_SUCCESS)
  {
    fprintf(stderr, "gearmand: gearman_queue_libpq_conf: %s\n",
            gearman_conf_error(&conf));
    return 1;
  }
#endif

  if (gearmand_protocol_http_conf(&conf) != GEARMAN_SUCCESS)
  {
    fprintf(stderr, "gearmand: gearman_protocol_http_conf: %s\n",
            gearman_conf_error(&conf));
    return 1;
  }

  /* Let gearman conf parse the command line arguments. */
  if (gearman_conf_parse_args(&conf, argc, argv) != GEARMAN_SUCCESS)
  {
    printf("\n%s\n\n", gearman_conf_error(&conf));
    printf("gearmand %s - %s\n\n", gearman_version(), gearman_bugreport());
    printf("usage: %s [OPTIONS]\n", argv[0]);
    gearman_conf_usage(&conf);
    return 1;
  }

  /* Check for option values that were given. */
  while (gearman_conf_module_value(&module, &name, &value))
  {
    if (!strcmp(name, "backlog"))
      backlog= atoi(value);
    else if (!strcmp(name, "daemon"))
    {
      switch (fork())
      {
      case -1:
        fprintf(stderr, "gearmand: fork:%d\n", errno);
        return 1;

      case 0:
        break;

      default:
        return 0;
      }

      if (setsid() == -1)
      {
        fprintf(stderr, "gearmand: setsid:%d\n", errno);
        return 1;
      }

      close_stdio= true;
    }
    else if (!strcmp(name, "file-descriptors"))
      fds= (rlim_t)atoi(value);
    else if (!strcmp(name, "help"))
    {
      printf("\ngearmand %s - %s\n\n", gearman_version(), gearman_bugreport());
      printf("usage: %s [OPTIONS]\n", argv[0]);
      gearman_conf_usage(&conf);
      return 1;
    }
    else if (!strcmp(name, "job-retries"))
      job_retries= (uint8_t)atoi(value);
    else if (!strcmp(name, "log-file"))
      log_info.file= value;
    else if (!strcmp(name, "listen"))
      host= value;
    else if (!strcmp(name, "port"))
      port= (in_port_t)atoi(value);
    else if (!strcmp(name, "pid-file"))
      pid_file= value;
    else if (!strcmp(name, "protocol"))
      continue;
    else if (!strcmp(name, "queue-type"))
      queue_type= value;
    else if (!strcmp(name, "threads"))
      threads= (uint32_t)atoi(value);
    else if (!strcmp(name, "user"))
      user= value;
    else if (!strcmp(name, "verbose"))
      verbose++;
    else if (!strcmp(name, "round-robin"))
      round_robin++;
    else if (!strcmp(name, "version"))
      printf("\ngearmand %s - %s\n", gearman_version(), gearman_bugreport());
    else if (!strcmp(name, "worker-wakeup"))
      worker_wakeup= (uint8_t)atoi(value);
    else
    {
      fprintf(stderr, "gearmand: Unknown option:%s\n", name);
      return 1;
    }
  }

  if (verbose == 0 && close_stdio)
  {
    /* If we can't remap stdio, it should not a fatal error. */
    fd = open("/dev/null", O_RDWR, 0);
    if (fd != -1)
    {
      if (dup2(fd, STDIN_FILENO) == -1)
      {
        fprintf(stderr, "gearmand: dup2:%d\n", errno);
        return 1;
      }

      if (dup2(fd, STDOUT_FILENO) == -1)
      {
        fprintf(stderr, "gearmand: dup2:%d\n", errno);
        return 1;
      }

      if (dup2(fd, STDERR_FILENO) == -1)
      {
        fprintf(stderr, "gearmand: dup2:%d\n", errno);
        return 1;
      }

      close(fd);
    }
  }

  if ((fds > 0 && _set_fdlimit(fds)) || _switch_user(user) || _set_signals())
    return 1;

  if (pid_file != NULL && _pid_write(pid_file))
    return 1;

  _gearmand= gearmand_create(host, port);
  if (_gearmand == NULL)
  {
    fprintf(stderr, "gearmand: Could not create gearmand library instance\n");
    return 1;
  }

  gearmand_set_backlog(_gearmand, backlog);
  gearmand_set_threads(_gearmand, threads);
  gearmand_set_job_retries(_gearmand, job_retries);
  gearmand_set_worker_wakeup(_gearmand, worker_wakeup);
  gearmand_set_log_fn(_gearmand, _log, &log_info, verbose);
  gearmand_set_round_robin(_gearmand, round_robin);

  if (queue_type != NULL)
  {
#ifdef HAVE_LIBDRIZZLE
    if (!strcmp(queue_type, "libdrizzle"))
    {
      ret= gearmand_queue_libdrizzle_init(_gearmand, &conf);
      if (ret != GEARMAN_SUCCESS)
        return 1;
    }
    else
#endif
#ifdef HAVE_LIBMEMCACHED
    if (!strcmp(queue_type, "libmemcached"))
    {
      ret= gearmand_queue_libmemcached_init(_gearmand, &conf);
      if (ret != GEARMAN_SUCCESS)
        return 1;
    }
    else
#endif
#ifdef HAVE_LIBSQLITE3
    if (!strcmp(queue_type, "libsqlite3"))
    {
      ret= gearmand_queue_libsqlite3_init(_gearmand, &conf);
      if (ret != GEARMAN_SUCCESS)
        return 1;
    }
    else
#endif
#ifdef HAVE_LIBPQ
    if (!strcmp(queue_type, "libpq"))
    {
      ret= gearmand_queue_libpq_init(_gearmand, &conf);
      if (ret != GEARMAN_SUCCESS)
        return 1;
    }
    else
#endif
#ifdef HAVE_LIBTOKYOCABINET
    if (!strcmp(queue_type, "libtokyocabinet"))
    {
      ret= gearmand_queue_libtokyocabinet_init(_gearmand, &conf);
      if (ret != GEARMAN_SUCCESS)
        return 1;
    }
    else
#endif        
    {
      fprintf(stderr, "gearmand: Unknown queue module: %s\n", queue_type);
      return 1;
    }
  }

  while (gearman_conf_module_value(&module, &name, &value))
  {
    if (strcmp(name, "protocol"))
      continue;

    if (!strcmp(value, "http"))
    {
      ret= gearmand_protocol_http_init(_gearmand, &conf);
      if (ret != GEARMAN_SUCCESS)
        return 1;
    }
    else
    {
      fprintf(stderr, "gearmand: Unknown protocol module: %s\n", value);
      return 1;
    }
  }

  ret= gearmand_run(_gearmand);

  if (queue_type != NULL)
  {
#ifdef HAVE_LIBDRIZZLE
    if (!strcmp(queue_type, "libdrizzle"))
      gearmand_queue_libdrizzle_deinit(_gearmand);
#endif
#ifdef HAVE_LIBMEMCACHED
    if (!strcmp(queue_type, "libmemcached"))
      gearmand_queue_libmemcached_deinit(_gearmand);
#endif
#ifdef HAVE_LIBSQLITE3
    if (!strcmp(queue_type, "libsqlite3"))
      gearmand_queue_libsqlite3_deinit(_gearmand);
#endif
#ifdef HAVE_LIBPQ
    if (!strcmp(queue_type, "libpq"))
      gearmand_queue_libpq_deinit(_gearmand);
#endif
#ifdef HAVE_LIBTOKYOCABINET
    if (!strcmp(queue_type, "libtokyocabinet"))
      gearmand_queue_libtokyocabinet_deinit(_gearmand);
#endif
  }

  while (gearman_conf_module_value(&module, &name, &value))
  {
    if (strcmp(name, "protocol"))
      continue;

    if (!strcmp(value, "http"))
      gearmand_protocol_http_deinit(_gearmand);
  }

  gearmand_free(_gearmand);

  if (pid_file != NULL)
    _pid_delete(pid_file);

  if (log_info.fd != -1)
    (void) close(log_info.fd);

  gearman_conf_free(&conf);

  return (ret == GEARMAN_SUCCESS || ret == GEARMAN_SHUTDOWN) ? 0 : 1;
}

static bool _set_fdlimit(rlim_t fds)
{
  struct rlimit rl;

  if (getrlimit(RLIMIT_NOFILE, &rl) == -1)
  {
    fprintf(stderr, "gearmand: Could not get file descriptor limit:%d\n",
            errno);
    return true;
  }

  rl.rlim_cur= fds;
  if (rl.rlim_max < rl.rlim_cur)
    rl.rlim_max= rl.rlim_cur;

  if (setrlimit(RLIMIT_NOFILE, &rl) == -1)
  {
    fprintf(stderr, "gearmand: Failed to set limit for the number of file "
                    "descriptors (%d). Try running as root or giving a "
                    "smaller value to -f.\n",
            errno);
    return true;
  }

  return false;
}

static bool _pid_write(const char *pid_file)
{
  FILE *f;

  f= fopen(pid_file, "w");
  if (f == NULL)
  {
    fprintf(stderr, "gearmand: Could not open pid file for writing: %s (%d)\n",
            pid_file, errno);
    return true;
  }

  fprintf(f, "%" PRId64 "\n", (int64_t)getpid());

  if (fclose(f) == -1)
  {
    fprintf(stderr, "gearmand: Could not close the pid file: %s (%d)\n",
            pid_file, errno);
    return true;
  }

  return false;
}

static void _pid_delete(const char *pid_file)
{
  if (unlink(pid_file) == -1)
  {
    fprintf(stderr, "gearmand: Could not remove the pid file: %s (%d)\n",
            pid_file, errno);
  }
}

static bool _switch_user(const char *user)
{
  struct passwd *pw;

  if (getuid() == 0 || geteuid() == 0)
  {
    if (user == NULL || user[0] == 0)
    {
      fprintf(stderr,
              "gearmand: Must specify '-u root' if you want to run as root\n");
      return true;
    }

    pw= getpwnam(user);
    if (pw == NULL)
    {
      fprintf(stderr, "gearmand: Could not find user '%s'\n", user);
      return 1;
    }

    if (setgid(pw->pw_gid) == -1 || setuid(pw->pw_uid) == -1)
    {
      fprintf(stderr, "gearmand: Could not switch to user '%s'\n", user);
      return 1;
    }
  }
  else if (user != NULL)
  {
    fprintf(stderr, "gearmand: Must be root to switch users\n");
    return true;
  }

  return false;
}

static bool _set_signals(void)
{
  struct sigaction sa;

  memset(&sa, 0, sizeof(struct sigaction));

  sa.sa_handler= SIG_IGN;
  if (sigemptyset(&sa.sa_mask) == -1 ||
      sigaction(SIGPIPE, &sa, 0) == -1)
  {
    fprintf(stderr, "gearmand: Could not set SIGPIPE handler (%d)\n", errno);
    return true;
  }

  sa.sa_handler= _shutdown_handler;
  if (sigaction(SIGTERM, &sa, 0) == -1)
  {
    fprintf(stderr, "gearmand: Could not set SIGTERM handler (%d)\n", errno);
    return true;
  }

  if (sigaction(SIGINT, &sa, 0) == -1)
  {
    fprintf(stderr, "gearmand: Could not set SIGINT handler (%d)\n", errno);
    return true;
  }

  if (sigaction(SIGUSR1, &sa, 0) == -1)
  {
    fprintf(stderr, "gearmand: Could not set SIGUSR1 handler (%d)\n", errno);
    return true;
  }

  return false;
}

static void _shutdown_handler(int signal_arg)
{
  if (signal_arg == SIGUSR1)
    gearmand_wakeup(_gearmand, GEARMAND_WAKEUP_SHUTDOWN_GRACEFUL);
  else
    gearmand_wakeup(_gearmand, GEARMAND_WAKEUP_SHUTDOWN);
}

static void _log(const char *line, gearman_verbose_t verbose, void *context)
{
  gearmand_log_info_st *log_info= (gearmand_log_info_st *)context;
  int fd;
  time_t t;
  char buffer[GEARMAN_MAX_ERROR_SIZE];

  if (log_info->file == NULL)
    fd= 1;
  else
  {
    t= time(NULL);

    if (log_info->fd != -1 && log_info->reopen < t)
    {
      (void) close(log_info->fd);
      log_info->fd= -1;
    }

    if (log_info->fd == -1)
    {
      log_info->fd= open(log_info->file, O_CREAT | O_WRONLY | O_APPEND, 0644);
      if (log_info->fd == -1)
      {
        fprintf(stderr, "gearmand: Could not open log file for writing (%d)\n",
                errno);
        return;
      }

      log_info->reopen= t + GEARMAND_LOG_REOPEN_TIME;
    }

    fd= log_info->fd;
  }

  snprintf(buffer, GEARMAN_MAX_ERROR_SIZE, "%5s %s\n",
           gearman_verbose_name(verbose), line);
  if (write(fd, buffer, strlen(buffer)) == -1)
    fprintf(stderr, "gearmand: Could not write to log file: %d\n", errno);
}
